### ISSUE TITLE
Question about the CSS loading strategy

### ISSUE THREAD

**Issue by bakura10** [Shopify/dawn#1392]
Hi,

Disclaimer: I am the developer of Maestrooo themes.

This is not rather a question, but more curiosity targeted to the dev team of Dawn and their own experiments. While splitting JS is a solved problem, CSS is a more complex solution and I am still not super convinced by the approach taken by Dawn (and why we are not doing it here) but I would be curious as I may be missing something here.

As of today, all our themes have a single CSS file (theme.css) containing the whole code. We try to minimize its size and for a full theme we manage to keep it around 35Kb.

Dawn, on the other hand, has a "base.css" file, and per-section (and per-component) CSS files. Generating this base and per-section CSS is not really a problem, but the loading strategy is.

In the case of a single, optimized CSS, loading in the <head>, the browser will block the rendering while downloading the file and render everything once done. This is optimal for UX: there is zero FOUT. With preloading the file can nearly be here instantaneously.

However, on Dawn, each section includes another blocking CSS in front of every sections: https://github.com/Shopify/dawn/blob/main/sections/main-list-collections.liquid#L1

As Shopify only allows two resources per page to be pre-loaded, those cannot.

As a consequence, the browser rendering is being blocked at the start of _every sections_.

From my understanding, this means that with a single pre-loaded file, the rendering may be blocked for 50ms (for instance) before being able to render the page content.

However, with Dawn strategy, the rendering will be blocked by 50ms (or maybe 40ms, because the "base.css" will be a bit smaller, but the improvement here will be negligeable), but then it will be blocked again at the start of the first section, and as the file will not be preloaded it will incur another 50ms, so way much more than what you would have with a single file.

I know there is a strategy to prevent CSS blocking by using a media print, but of course this is the worse approach as it cause a very bad FOUT.

Am I missing something here, or may the approach taken by Dawn effectively suboptimal?

**Comment by peristylethemes** [comment]
Hi @[bakura10](https://github.com/bakura10),
I have tested both single CSS file approach and with Dawn's approach. I am seeing a huge performance boost in case of a single CSS file. When I use modular CSS in every sections, lighthouse score drops (at least 10 points) and render blocking is reported. Am I correct?

**Comment by bakura10** [comment]
I don't think Dawn approach is the best one when it comes to CSS. The biggest issue actually is more the fact that when you load a CSS it is re-parsed as many times as it is included. If you add several instances of a section that load a CSS, then this CSS will appear duplicated in the inspectors.

With using one CSS file, especially if this one is pretty lightweight and preloaded, I think this offers better performance and dev experience.

**Comment by bakura10** [comment]
Keeping a whole CSS within 10Kb is really hard if you have more than a few sections. One approach you can try is using Tailwind to keep your CSS super light. It is a different approach of CSS, but can lead to very lightweight stylesheet ;)

**Comment by tin-soldier** [comment]
Have tested combining all css files except those in <noscript> tags or those starting with 'template'. The Lighthouse score went from 88 to 95.

**Comment by bakura10** [comment]
Thanks for doing the test, it does not surprise me. I’d say that as long the CSS is less than 35-40Kb, splitting has no benefit.

I also went backward on JS: I realised that splitting file actually resulted in more JS being loaded due to poorer usage of Gzip compression. The same will likely apply to CSS.

**Comment by kyle-tully** [comment]
Found this thread while doing a speed optimisation project on a Dawn theme that has loading issues because of this very reason.

**Comment by kyle-tully** [comment]
> Did anyone test inlining the css in the sections rather than linking? and the effect that had? Is that a viable strategy?

Yes I inlined one or two critical sections and it helped.

**Comment by vfonic** [comment]
I think this is a very important question. I'm searching for a speed-optimized theme to recommend to a client and this is the only reason why I'm hesitant to recommend Dawn.

Can someone from Shopify provide some insight?

ping @eugenekasimov @sofiamatulis @kmeleta 
Thank you! 🙏

**Comment by bakura10** [comment]
@vfonic , while Dawn may not have taken the best approach here and that there are slightly faster way to do it, it still remains a very fast theme. Any themes from the theme store or Dawn will never be a bottleneck and all are fast enough by default so that they do not cause issue. We are talking about a few ms of potential difference here. Any apps, large image… will have a much larger impact.

**Comment by jeremyquinton** [comment]
I'm very perplexed by the dawn them. A basic page has over 100 requests for css/javascript and image files etc. It's not very performant on slower mobile connections and ranks fairly poorly in google pagespeed insights. 

Is there any solution at all to reducing the amount of http requests that are needed. 
As an example main-product.liquid below has 8 css files. 

  ```{{ 'section-main-product.css' | asset_url | stylesheet_tag }}
  {{ 'component-accordion.css' | asset_url | stylesheet_tag }}
  {{ 'component-price.css' | asset_url | stylesheet_tag }}
  {{ 'component-rte.css' | asset_url | stylesheet_tag }}
  {{ 'component-slider.css' | asset_url | stylesheet_tag }}
  {{ 'component-rating.css' | asset_url | stylesheet_tag }}
  {{ 'component-loading-overlay.css' | asset_url | stylesheet_tag }}
  {{ 'component-deferred-media.css' | asset_url | stylesheet_tag }}```
  
One thing that I could do here is remove all css files and make the css inline or potentially combine all 8 files into one taking 8 http requests to 1. This would cause problems if css is needed to be edited but in most instances I'm not touching any of the css for development anyways. 

Can anyone suggest a them that is super performant aka doesn't have so many requests per page.
